### PR TITLE
Improve Javascript state serialization

### PR DESCRIPTION
**Pull request by ruwanta**
### Proposed changes in this pull request

- Enable javascript context to be moved across the cluster nodes by
allowing serialization on selected context variables.

- Removed Javascript Cache.

### When should this PR be merged

After proper testing

### Follow up actions

- Test on dual node cluster with multi-step authentication

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/java/org/wso2/carbon/identity/application/authentication/framework/MockUiAuthenticator.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+    public MockUiAuthenticator(String name, SubjectCallback subjectCallback) {
+        this.name = name;
+        this.subjectCallback = subjectCallback;
+    }
+
+    @Override
+    protected void initiateAuthenticationRequest(HttpServletRequest request, HttpServletResponse response,
+            AuthenticationContext context) throws AuthenticationFailedException {
+        super.initiateAuthenticationRequest(request, response, context);
```
Comment (pulasthi7) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/java/org/wso2/carbon/identity/application/authentication/framework/MockUiAuthenticator.java]: Shall we add the empty lines after the method signature to make it consistent with others?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsGraphBuilder.java [hunk 2]
```diff
@@ -220,46 +210,46 @@ public void executeStepInAsyncEvent(Map<String, Object> parameterMap) {
...
-        } else {
-            source = String.valueOf(value);
-        }
+    private static DynamicDecisionNode createDynamicDecisionNode(Map<String, Object> eventsMap) {
 
-        return new JsBasedEvaluator(source, isFunction);
+        DynamicDecisionNode decisionNode = new DynamicDecisionNode();
+        eventsMap.forEach((key, value) -> {
+            if (value instanceof ScriptObjectMirror) {
+                SerializableJsFunction fn = SerializableJsFunction.toSerializableForm(key, (ScriptObjectMirror) value);
```
Comment (emswbandara) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsGraphBuilder.java]: better to add a descriptive variable name

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/GraphBasedSequenceHandlerClusterTest.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        SequenceConfig sequenceConfig = configurationLoader
+                .getSequenceConfig(context, Collections.<String, String[]>emptyMap(), sp1);
+        context.setSequenceConfig(sequenceConfig);
+        return context;
+    }
+
+    @DataProvider(name = "roleBasedDataProvider")
+    public Object[][] getRoleBasedUserRolesData() {
+        return new Object[][] {
+                //                {"js-sp-2.xml",  1, false},
```
Comment (emswbandara) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/GraphBasedSequenceHandlerClusterTest.java]: remove commented out code.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/resources/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/js-sp-3.xml [hunk 4]
```diff
@@ -0,0 +1,115 @@
+<!--
+  ~ Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (emswbandara) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/resources/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/js-sp-3.xml]: change license year

**Pull request merged**